xds/internal/balancer/gracefulswitch/gracefulswitchbalancer.go

```diff
@@ -0,0 +1,398 @@
+	// induce deadlock because it can callback into a function that also
+	// requires gsb.mu. However, in between reading into a local var pointer and
+	// updating that local var pointer, the current balancer can potentially be
+	// closed in an updateState call. The reason the pending cannot also be
+	// closed is we are guaranteed that grpc will call the balancer API
+	// synchronously, and for pending to be deleted would require another
+	// SwitchTo(), which is outside of the state space of the potential
+	// scenarios within the context of this single UpdateClientConnState() call.
+	// However, in the scenario that the current can possibly be deleted
+	// (current + pending balancers are both populated), this Update will always
+	// go to the pending. Thus, there is a guarantee that we never break the API
+	// guarantee for the current balancer by closing before updating.
+	balToUpdate.ResolverError(err)
+}
+
+func (gsb *gracefulSwitchBalancer) UpdateSubConnState(sc balancer.SubConn, state balancer.SubConnState) {
+	if gsb.closed.HasFired() {
+		// Logger
+		return
+	}
+	// At any given time, the current balancer may be closed in the situation
+	// where there is a current balancer and a pending balancer. Thus, the
+	// swap() (close() + clearing of SubCons) operation is also guarded by this
+	// currentMu. Whether the current has been deleted or not will be checked
+	// from reading from the scToSubBalancer map, as that will be cleared in the
+	// atomic swap() operation. We are guaranteed that there can only be one
+	// balancer closed within this UpdateSubConnState() call, as multiple
+	// balancers closed would require another call to UpdateClientConnState()
+	// which we are guaranteed won't be called concurrently.
+	// UpdateSubConnState() is different to UpdateClientConnState() and
+	// ResolverError(), as for those functions you don't update the current
+	// balancer in the situation where the current balancer could be closed
+	// (current + pending balancer populated).
+	gsb.currentMu.Lock()
+	defer gsb.currentMu.Unlock()
+	gsb.mu.Lock()
+	// This SubConn update will forward to the current balancer even if there is
+	// a pending present. This is because if this balancer does not forward the
+	// update, the picker from the current will not be updated with SubConns,
+	// leading to the possibility that the ClientConn constantly picks bad
+	// SubConns in the Graceful Switch period.
+	balToUpdate, ok := gsb.scToSubBalancer[sc]
+	if !ok {
+		gsb.mu.Unlock()
+		return
+	}
+	if state.ConnectivityState == connectivity.Shutdown {
+		delete(gsb.scToSubBalancer, sc)
+	}
+	gsb.mu.Unlock()
+	balToUpdate.UpdateSubConnState(sc, state)
+}
+
+func (gsb *gracefulSwitchBalancer) Close() {
+	gsb.closed.Fire()
+	gsb.mu.Lock()
+	for sc := range gsb.scToSubBalancer {
+		gsb.cc.RemoveSubConn(sc)
+		delete(gsb.scToSubBalancer, sc)
+	}
+	currentBalancerToUpdate := gsb.balancerCurrent
+	if gsb.balancerCurrent != nil {
```

Comment:
Unconditional assignment is always better than test + assign.  "Branches are bad".